cli/repair/patch.go

```diff
@@ -0,0 +1,174 @@
+package repair
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/tarantool/cartridge-cli/cli/common"
+	"github.com/tarantool/cartridge-cli/cli/context"
+	"github.com/tarantool/cartridge-cli/cli/project"
+	"github.com/tarantool/cartridge-cli/cli/templates"
+)
+
+const (
+	confapplierWishStateTimeout = 10
+)
+
+func patchConf(patchFunc PatchConfFuncType, topologyConf *TopologyConfType, ctx *context.Ctx) ([]common.ResultMessage, error) {
+	var resMessages []common.ResultMessage
+
+	currentConfContent, err := topologyConf.MarshalContent()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to marshal current content: %s", err)
+	}
+
+	if err := patchFunc(topologyConf, ctx); err != nil {
+		return nil, fmt.Errorf("Failed to patch topology config: %s", err)
+	}
+
+	newConfContent, err := topologyConf.MarshalContent()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get new config content: %s", err)
+	}
+
+	if ctx.Repair.DryRun || ctx.Cli.Verbose {
+		configDiff, err := getDiffLines(currentConfContent, newConfContent, "", "")
+		if err != nil {
+			return nil, fmt.Errorf("Failed to get config difference: %s", err)
+		}
+
+		if len(configDiff) > 0 {
+			resMessages = append(resMessages, common.GetInfoMessage((strings.Join(configDiff, "\n") + "\n")))
+		} else {
+			resMessages = append(resMessages, common.GetInfoMessage("Topology config wasn't changed"))
+		}
+	}
+
+	return resMessages, nil
+}
+
+func rewriteConf(topologyConfPath string, topologyConf *TopologyConfType) ([]common.ResultMessage, error) {
+	var resMessages []common.ResultMessage
+
+	resMessages = append(resMessages, common.GetDebugMessage("Topology config file: %s", topologyConfPath))
+
+	backupPath, err := createFileBackup(topologyConfPath)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create topology config backup: %s", err)
+	}
+	resMessages = append(resMessages, common.GetDebugMessage("Created backup file: %s", backupPath))
+
+	newConfContent, err := topologyConf.MarshalContent()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get new config content: %s", err)
+	}
+
+	confFile, err := os.OpenFile(topologyConfPath, os.O_WRONLY|os.O_TRUNC, 0755)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to open a new config: %s", err)
+	}
+
+	if _, err := confFile.Write(newConfContent); err != nil {
+		return nil, fmt.Errorf("Failed to write a new config: %s", err)
+	}
+
+	return resMessages, nil
+}
+
+func reloadConf(topologyConfPath string, instanceName string, ctx *context.Ctx) ([]common.ResultMessage, error) {
```

Comment:
I hope we could move this function to separate file.

May be in future we will implement something like `cartridge reload` or it could be useful for https://github.com/tarantool/cartridge-cli/issues/249

Reply:
Yes, but it seems that this function should be modified for this (or only some part will be specified).
I suggest moving this function later, when we understood, where it can be used,

Reply:
Well, ok